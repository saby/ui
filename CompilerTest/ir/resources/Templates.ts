import * as Element from 'wml!CompilerTest/ir/resources/Element';
import * as If from 'wml!CompilerTest/ir/resources/If';
import * as For from 'wml!CompilerTest/ir/resources/For';
import * as Foreach from 'wml!CompilerTest/ir/resources/Foreach';
import * as InlineTemplate from 'wml!CompilerTest/ir/resources/InlineTemplate';
import * as ContentOption from 'wml!CompilerTest/ir/resources/ContentOption';
import * as StaticPartial from 'wml!CompilerTest/ir/resources/StaticPartial';
import * as DynamicPartial from 'wml!CompilerTest/ir/resources/DynamicPartial';
import * as DynamicInline from 'wml!CompilerTest/ir/resources/DynamicInline';

export default {
    Element,
    If,
    For,
    Foreach,
    InlineTemplate,
    ContentOption,
    StaticPartial,
    DynamicPartial,
    DynamicInline
};
