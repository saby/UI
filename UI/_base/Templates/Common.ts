import { IoC } from 'Env/Env';

function logError(tag, err) {
   var logger = IoC.resolve('ILogger') || console;
   logger.error(tag, err.message, err);
}

export {
   logError
}
