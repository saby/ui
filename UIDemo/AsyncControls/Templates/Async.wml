<div class="asyncDemo__column">
    <Controls.Container.Async templateName="Controls/heading:Title">
        <ws:partial template="{{ content.resolvedTemplate }}" caption="Async" readOnly="{{true}}" fontColorStyle="secondary" fontSize="l"/>
    </Controls.Container.Async>
    <div class="asyncDemo__column">
        <ws:for data="i in 3">
            <Controls.Container.Async templateName="Controls/input:Number" attr:class="demo-Async__input_{{i}}">
                <ws:partial template="{{ content.resolvedTemplate }}" value="{{i}}" />
            </Controls.Container.Async>
        </ws:for>
    </div>
</div>
