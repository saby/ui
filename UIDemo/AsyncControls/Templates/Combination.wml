<div class="asyncDemo__column">
    <Controls.Container.Async templateName="Controls/heading:Title">
        <ws:partial template="{{ content.resolvedTemplate }}" caption="Combination {{_options.indexOpt}}" readOnly="{{true}}" fontColorStyle="secondary" fontSize="l"/>
    </Controls.Container.Async>

    <ws:for data="i in 2">
        <Controls.input:Text  value="not async {{i}}" attr:class="demo-Combination__input_{{i}}"/>
        <Controls.Container.Async templateName="Controls/input:Text">
            <ws:partial template="{{ content.resolvedTemplate }}" value="async {{i}}" />
        </Controls.Container.Async>
    </ws:for>
</div>
