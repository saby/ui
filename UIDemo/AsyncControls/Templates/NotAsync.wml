<div class="asyncDemo__column">
    <Controls.heading:Title  caption="Not Async" readOnly="{{true}}" fontColorStyle="secondary" fontSize="l"/>
    <div class="asyncDemo__column">
        <ws:for data="i in 3">
            <Controls.input:Number attr:class="demo-NotAsync__input_{{i}}" value="{{i}}"/>
        </ws:for>
    </div>
</div>
