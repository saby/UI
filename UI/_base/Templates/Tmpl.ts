import { logError } from "./Common";
import { TClosure } from 'UI/Executor';
import { _FocusAttrs } from 'UI/Focus';

function createLostFunction(err, ext) {
   logError(ext + '!', err);
   var wrapper = function () {
      return '<div>' + err.message + '</div>';
   };
   // @ts-ignore
   wrapper.stable = true;
   // @ts-ignore
   wrapper.includedFunctions = {};
   return wrapper;
}
function setToJsonForFunction(func, moduleName, path = undefined) {
   func.toJSON = function() {
      var serialized = {
         $serialized$: 'func',
         module: moduleName
      };
      if (path) {
         // @ts-ignore
         serialized.path = path;
      }
      return serialized;
   };
}
function resolverControls(path) {
   return 'tmpl!' + path;
}
function createTemplate(name, html, tmpl, conf, load) {
   try {
      tmpl.template(html, resolverControls, conf).handle(function (traversed) {
         try {
            var templateFunction = tmpl.func(traversed, conf);
            Object.keys(templateFunction.includedFunctions).forEach(function(elem) {
               setToJsonForFunction(templateFunction.includedFunctions[elem], 'tmpl!' + name, 'includedFunctions.' + elem);
            });
            // Чтобы отличать функции старого шаблонизатора от нового
            templateFunction.stable = true;
            var closured = function () {
               return templateFunction.apply(this, tmpl.addArgument(TClosure, arguments));
            };
            Object.defineProperty(closured, 'name', { 'value': templateFunction.name, configurable: true });
            // @ts-ignore
            closured.stable = true;
            // @ts-ignore
            closured.includedFunctions = templateFunction.includedFunctions;
            setToJsonForFunction(closured, 'tmpl!' + name);

            // @ts-ignore
            closured.reactiveProps = traversed.reactiveProps;
            // @ts-ignore
            closured.originalFunction = templateFunction;
            // @ts-ignore
            closured.__beforeMount = _FocusAttrs.prepareAttrsForFocus;

            load(closured);
            load = undefined;
         } catch (err) {
            err.message = 'Error while traversing template "' + name + '": ' + err.message;
            load(createLostFunction(err, 'tmpl'));
            load = undefined;
         }
      }, function (err) {
         err.message = 'Error while creating template "' + name + '": ' + err.message;
         load(createLostFunction(err,' tmpl'));
         load = undefined;
      });
   } catch (err) {
      err.message = 'Error while parsing template "' + name + '": ' + err.message;
      load(createLostFunction(err, 'tmpl'));
      load = undefined;
   }
}
export { createTemplate }
