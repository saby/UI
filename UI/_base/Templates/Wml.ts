import {logError} from "./Common";
import {_FocusAttrs} from 'UI/Focus';

var global = (function() {
   return this || (0, eval)('this');
}());

function showAlertOnTimeoutInBrowser(err) {
   if (!err) { return false; }
   // @ts-ignore
   if (showAlertOnTimeoutInBrowser.isFired) { return false; }
   var REQUIRE_TIMEOUT_TYPE = 'timeout'
   if (err.requireType !== REQUIRE_TIMEOUT_TYPE) { return false; }
   if (typeof window === 'undefined') { return false; }
   if (global.wsConfig && global.wsConfig.showAlertOnTimeoutInBrowser === false) { return false; }
   var importantModules = err.requireModules.map(function (moduleName) {
      return moduleName.substr(0, 4) !== 'css!';
   });
   if (importantModules.length === 0) { return false; }
   alert('Произошла ошибка загрузки ресурса. Проверьте интернет соединение и повторите попытку.');
   // @ts-ignore
   showAlertOnTimeoutInBrowser.isFired = true;
   throw err;
}
function loadModule(name, file, load) {
   try {
      eval(file);
      requirejs(['wml!' + name], function(module) {
         module.__beforeMount = _FocusAttrs.prepareAttrsForFocus;
         load(module);
      }, function(err) {
         logError('Wml module loading error', err);
         load.error(err);
      });
   } catch(err) {
      logError('Wml module executing error', err);
      load.error(err);
   }
}
function createTemplate(name, html, tmpl, conf, load, ext) {
   try {
      if (!conf.fileName) {
         conf.fileName = name;
      }
      tmpl.getFile(html, conf, function (file) {
         loadModule(name, file, load);
      }, function (err) {
         err.message = 'Error while parsing template "' + name + '": ' + err.message;
         try {
            var timeoutAlert = showAlertOnTimeoutInBrowser(err);
            if (!timeoutAlert) {
               logError('Template', err);
            }
         } catch (err) {
            logError('Template', err);
         }
         load.error(err);
      }, ext);
   } catch (err) {
      err.message = 'Error while parsing template "' + name + '": ' + err.message;
      logError('Template', err);
      load.error(err);
   }
}

export {
   createTemplate,
   loadModule
}
