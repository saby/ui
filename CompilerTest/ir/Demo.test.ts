/* eslint-disable @typescript-eslint/no-magic-numbers,@typescript-eslint/no-shadow */

import Templates from './resources/Templates';

function invoke(fn: Function, data: object = {}): unknown[] {
    return fn(data, { }, { }, false);
}

describe('Compiler/IR', () => {
    it('should generate root tag', () => {
        expect(invoke(Templates.Element)).toMatchSnapshot();
    });

    describe('test conditional chain', function () {
        it('should generate body of if node', () => {
            const data = {
                ifCondition: true,
                elifCondition: false,
                userName: 'User'
            };

            expect(invoke(Templates.If, data)).toMatchSnapshot();
        });
        it('should generate body of elif node', () => {
            const data = {
                ifCondition: false,
                elifCondition: true,
                userName: 'User'
            };

            expect(invoke(Templates.If, data)).toMatchSnapshot();
        });
        it('should generate body of else node', () => {
            const data = {
                ifCondition: false,
                elifCondition: false,
                userName: 'User'
            };

            expect(invoke(Templates.If, data)).toMatchSnapshot();
        });
    });

    it('should iterate collection via for method', () => {
        const data = {
            userName: 'User',
            iterator: {
                init() {
                    this.index = 0;
                },
                test() {
                    return this.index < 3;
                },
                update() {
                    this.index++;
                },
                getText() {
                    return `Content of ${this.index} item`;
                }
            }
        };

        expect(invoke(Templates.For, data)).toMatchSnapshot();
    });
    it('should iterate collection via foreach method', () => {
        const data = {
            array: [111, 222, 333]
        };

        expect(invoke(Templates.Foreach, data)).toMatchSnapshot();
    });
    it('should generate inline template', () => {
        expect(invoke(Templates.InlineTemplate)).toMatchSnapshot();
    });
    it('should generate content option', () => {
        expect(invoke(Templates.ContentOption)).toMatchSnapshot();
    });
    it('should generate static partial template', () => {
        expect(invoke(Templates.StaticPartial)).toMatchSnapshot();
    });
    it('should generate dynamic partial template', () => {
        const data = {
            tmpl: Templates.Element
        };

        expect(invoke(Templates.StaticPartial, data)).toMatchSnapshot();
    });
    it('should generate dynamic inline template', () => {
        const data = {
            tmpl: 'inline'
        };

        expect(invoke(Templates.DynamicInline, data)).toMatchSnapshot();
    });
});
