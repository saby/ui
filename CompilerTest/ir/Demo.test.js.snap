// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler/IR should generate content option 1`] = `"<div class=\\"root\\"><div class=\\"inline\\"><span class=\\"content\\">Hello, Content option!</span></div></div>"`;

exports[`Compiler/IR should generate dynamic inline template 1`] = `"<div class=\\"root\\"><div class=\\"inline\\">Text</div></div>"`;

exports[`Compiler/IR should generate dynamic partial template 1`] = `"<div class=\\"root\\"><div class=\\"custom-class\\">Hello, World!</div></div>"`;

exports[`Compiler/IR should generate inline template 1`] = `"<div class=\\"root\\"><div class=\\"inline\\">Text</div></div>"`;

exports[`Compiler/IR should generate root tag 1`] = `"<div class=\\"custom-class\\">Hello, World!</div>"`;

exports[`Compiler/IR should generate static partial template 1`] = `"<div class=\\"root\\"><div class=\\"custom-class\\">Hello, World!</div></div>"`;

exports[`Compiler/IR should iterate collection via for method 1`] = `"<div class=\\"root\\"><div class=\\"item\\">Content of 0 item</div><div class=\\"item\\">Content of 1 item</div><div class=\\"item\\">Content of 2 item</div></div>"`;

exports[`Compiler/IR should iterate collection via foreach method 1`] = `"<div class=\\"root\\"><div class=\\"item\\">array[0] = 111</div><div class=\\"item\\">array[1] = 222</div><div class=\\"item\\">array[2] = 333</div></div>"`;

exports[`Compiler/IR test conditional chain should generate body of elif node 1`] = `"<div class=\\"elif-branch\\">Hello, User! It's 3rd expression</div>"`;

exports[`Compiler/IR test conditional chain should generate body of else node 1`] = `"<div class=\\"else-branch\\">Hello, User! It's 4th expression</div>"`;

exports[`Compiler/IR test conditional chain should generate body of if node 1`] = `"<div class=\\"if-branch\\">Hello, User! It's 2nd expression</div>"`;
